### PR TITLE
issue 4585 - backend redesign phase 3c - dbregion test removal

### PR DESCRIPTION
**Pull request by progier389**
Change description:
 [A} fix the slapd_pr_strerror to handle dbimpl error code (and modify dblayer_strerror to use the strings from the generated include file)
 [B] removes the test about bdb region files used to determine if import or reindex task are in progress. ( File locks are used instead. )
  Note: lock files descriptor are stored in the ldbm_instance while the db instance is open
  and most of the logic is handled in dblayer.c 

Added a test to run concurrently 2 "jobs" (i.e online and offline commands about import or export suffix1 or suffix2, reindex suffix1 or suffix2, backup or restore or do nothing. 
I run first the test on   build 1.4.2.12 and generated a reference file, then compared the result with current version
I realized that I got same result with and without the file lock checking so I simplified thing and removed the lock (by reverting dblayer.c to master's version)

So now there is only the slapd_pr_strerror change
 Note: The fix is quite small (except the test case) most of the changes are due to the fact that "char *msg" is now "const char *msg" 

relates:
#4469 - Backend redesign phase 3a - (dbimpl API implementation + bdb dependencies removal in back-ldbm)
#4552 - Backend redesign phase 3b - bdb dependencies removal in replication plugins
#4585 - Backend redesign phase 3c: remove the dbregion test

Reviewed by: 

Platforms tested: F32

### CODE REVIEW

File: ldap/servers/slapd/back-ldbm/dblayer.c [hunk 1]
```diff
@@ -270,6 +277,117 @@ dblayer_start(struct ldbminfo *li, int dbmode)
...
+        slapi_log_err(SLAPI_LOG_ERR, "dblayer_import_lock",
+                      "Failed to get import lock filename\n");
+    if (pt != filename) {
+        slapi_ch_free_string(&pt);
+    }
+    if (!pt) {
+        slapi_log_err(SLAPI_LOG_ERR, "dblayer_import_lock",
+                      "Failed to get import lock filename\n");
+        return SLAPI_FAIL_GENERAL;
+    }
```
Comment (mreynolds389) [inline ldap/servers/slapd/back-ldbm/dblayer.c]: If pt != filename, then slapi_ch_free_string() sets pt to NULL, then the pt NULL check will log an error and exit.  Is that indented?  If it is, then you could combine these two code blocks.

File: ldap/servers/slapd/back-ldbm/dblayer.c [hunk 2]
```diff
@@ -270,6 +277,117 @@ dblayer_start(struct ldbminfo *li, int dbmode)
...
+            fd = inst->import_lock_fd;
+            if (fd>=0) {
+               /* Let's unlock other processes */
+               unlink(filename);
+               /* Let's unlock other threads */
+               inst->import_lock_fd = -1;
+               /* Lets release the resources */
+               flock(fd, LOCK_UN);
+               close(fd);
+            }
```
Comment (mreynolds389) [inline ldap/servers/slapd/back-ldbm/dblayer.c]: indentation is off in the above code block.  :-)  There are more to come, sorry...

File: ldap/servers/slapd/back-ldbm/dblayer.c [hunk 3]
```diff
@@ -270,6 +277,117 @@ dblayer_start(struct ldbminfo *li, int dbmode)
...
+            return 0;
+
+        case LOCK_SET:
+            fd = open(filename, O_CREAT | O_CLOEXEC | O_RDONLY, 0440);
+            if (fd < 0) {
+                slapi_log_err(SLAPI_LOG_ERR, "dblayer_import_lock",
+                      "Failed to open import lock file %s errno=%d\n",
+                      filename, errno);
+                return SLAPI_FAIL_GENERAL;
```
Comment (mreynolds389) [inline ldap/servers/slapd/back-ldbm/dblayer.c]: indentation #1

File: ldap/servers/slapd/back-ldbm/dblayer.c [hunk 4]
```diff
@@ -270,6 +277,117 @@ dblayer_start(struct ldbminfo *li, int dbmode)
...
+                close(fd);
+                return LDAP_BUSY;
+            }
+            if (rc) {
+                close(fd);
+                return LDAP_BUSY;
+                slapi_log_err(SLAPI_LOG_ERR, "dblayer_import_lock",
+                      "Failed to set lock on file %s errno=%d\n",
+                      filename, errno);
+                return SLAPI_FAIL_GENERAL;
```
Comment (mreynolds389) [inline ldap/servers/slapd/back-ldbm/dblayer.c]: indentation #2

File: ldap/servers/slapd/back-ldbm/dblayer.c [hunk 5]
```diff
@@ -270,6 +277,117 @@ dblayer_start(struct ldbminfo *li, int dbmode)
...
+            }
+            rc = flock(fd, LOCK_NB|LOCK_EX);
+            if (rc && errno == EWOULDBLOCK) {
+                close(fd);
+                return LDAP_BUSY;
+            }
+            if (rc) {
+                close(fd);
+                return LDAP_BUSY;
```
Comment (mreynolds389) [inline ldap/servers/slapd/back-ldbm/dblayer.c:348]: There are two "returns" in this code block.  Is it LDAP_BUSY or SLAPI_GENERAL_FAIL

File: ldap/servers/slapd/back-ldbm/dblayer.c [hunk 6]
```diff
@@ -270,6 +277,117 @@ dblayer_start(struct ldbminfo *li, int dbmode)
...
+            if (rc == 0) {
+                flock(fd, LOCK_NB|LOCK_UN);
+                close(fd);
+                return 0;
+            }
+            close(fd);
+            if (errno != EWOULDBLOCK) {
+                slapi_log_err(SLAPI_LOG_ERR, "dblayer_import_lock",
+                      "Failed to set lock on file %s errno=%d\n",
+                      filename, errno);
```
Comment (mreynolds389) [inline ldap/servers/slapd/back-ldbm/dblayer.c]: indentation #4

File: ldap/servers/slapd/back-ldbm/dblayer.c [hunk 7]
```diff
@@ -270,6 +277,117 @@ dblayer_start(struct ldbminfo *li, int dbmode)
...
+        case LOCK_TEST:
+            fd = open(filename, O_CLOEXEC | O_RDONLY, 0440);
+            if (fd < 0) {
+                if (mode == LOCK_TEST && errno == ENOENT) {
+                    /* file does not exists ==> there is no lock */
+                    return 0;
+                }
+                slapi_log_err(SLAPI_LOG_ERR, "dblayer_import_lock",
+                      "Failed to open import lock file %s errno=%d\n",
+                      filename, errno);
```
Comment (mreynolds389) [inline ldap/servers/slapd/back-ldbm/dblayer.c]: indentation #3

File: ldap/servers/slapd/back-ldbm/dblayer.c (hunk 4 above)
Comment (mreynolds389) [reply ldap/servers/slapd/back-ldbm/dblayer.c]: There are also two "returns" in this code block

**Comment by mreynolds389** [comment]
LGTM

**Review by Firstyear — APPROVED** [review]

**Comment by Firstyear** [comment]
Looks good from me too

File: ldap/servers/slapd/back-ldbm/dbimpl.c [hunk 8]
```diff
@@ -298,33 +298,30 @@ int dblayer_set_dup_cmp_fn(Slapi_Backend *be, struct attrinfo *a, dbi_dup_cmp_t
...
+     * Cannot use slapi_pr_strerror() because dbscan is not linked with
+     * libslapd so lets use dberrstrs.h
+     * And table is small enough that it is not worth to use bsearch;
+     */
+    static const struct {
+        int errcode;
+        const char *errmsg;
+    } errtab[] = {
+#include "dberrstrs.h"
+        { DBI_RC_SUCCESS, "No error." }
```
Comment (tbordaz) [inline ldap/servers/slapd/back-ldbm/dbimpl.c:314]: If dbimpl.h is looking like below, could we remove the special handling of DBI_RC_SUCCESS ?
<pre>
typedef enum {
    DBI_RC_SUCCESS,               /* 0, Database operation error: No error */
    DBI_RC_UNSUPPORTED = -0x3200, /* -12800, Database operation error: Operation not supported. */
    ...
</pre>

File: ldap/servers/slapd/back-ldbm/db-bdb/bdb_layer.c [hunk 9]
```diff
@@ -2017,7 +2017,7 @@ bdb_pre_close(struct ldbminfo *li)
     conf = (bdb_config *)li->li_dblayer_config;
     bdb_db_env *pEnv = (bdb_db_env *)priv->dblayer_env;
 
-    if (conf->bdb_stop_threads) /* already stopped.  do nothing... */
+    if (conf->bdb_stop_threads || !pEnv) /* already stopped.  do nothing... */
```
Comment (tbordaz) [inline ldap/servers/slapd/back-ldbm/db-bdb/bdb_layer.c:2020]: Would you add a comment in which case 'pEnv' can be NULL ?

**Review by tbordaz — APPROVED** [review]
LGTM

**Pull request merged**